sys/net/transport_layer/ng_udp/ng_udp.c

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright (C) 2015 Freie Universität Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @{
+ * @ingroup     sys_net_udp
+ *
+ * @file
+ * @brief       UDP implementation
+ *
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ * @}
+ */
+
+#include <stdint.h>
+
+#include "kernel.h"
+#include "msg.h"
+#include "thread.h"
+#include "net/ng_udp.h"
+#include "net/ng_nettype.h"
+#include "net/ng_netapi.h"
+#include "net/ng_netreg.h"
+#include "net/ng_pktbuf.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+/**
+ * @brief   Save the UDP's thread PID for later reference
+ */
+static kernel_pid_t _pid = KERNEL_PID_UNDEF;
+
+/**
+ * @brief   Allocate memory for the UDP thread's stack
+ */
+static char _stack[NG_UDP_STACK_SIZE];
+
+/**
+ * @brief   Calculate the UDP checksum dependent on the network protocol
+ *
+ * @param[in] pkt       pointer to the packet in the packet buffer
+ *
+ * @return              the checksum of the pkt
+ * @return              0 on error
+ */
+uint16_t _calc_checksum(pktsnip_t *pkt)
+{
+    uint16_t cksum = 0;
+    /* TODO: calculate checksum depending on used IP version */
+    return cksum;
+}
+
+void _receive(pktsnip_t *pkt)
+{
+    ng_pktsnip_t *payload;
+    ng_pktsnip_t *udp = pkt;
+    ng_udp_hdr_t *hdr;
+    msg_t msg;
+    ng_netreg_entry_t *reciever;
+
+    /* mark UDP header */
+    while (udp && udp->type != NG_NETTYPE_UDP) {
```

Comment:
(an even saver approach, that doesn't need a preceiding header do anything with its payload would be to find the last `pktsnip_t` in the packet [which must be the payload]):

LL_SEARCH_SCALAR(pkt, payload, next, NULL);
// no abort needed ;-)
pkt = ng_pktbuf_start_write(pkt);